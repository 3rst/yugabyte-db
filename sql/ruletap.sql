\unset ECHO
\i test_setup.sql

SELECT plan(66);
--SELECT * FROM no_plan();

-- This will be rolled back. :-)
SET client_min_messages = warning;
CREATE TABLE public.sometab(
    id    INT NOT NULL PRIMARY KEY,
    name  TEXT DEFAULT '',
    numb  NUMERIC(10, 2),
    myint NUMERIC(8)
);
CREATE RULE ins_me AS ON INSERT TO public.sometab DO NOTHING;
CREATE RULE upd_me AS ON UPDATE TO public.sometab DO ALSO SELECT now();

CREATE TABLE public.toview ( id INT );
CREATE RULE "_RETURN" AS ON SELECT TO public.toview DO INSTEAD SELECT 42 AS id;

RESET client_min_messages;

/****************************************************************************/
-- Test has_rule() and hasnt_rule().

SELECT * FROM check_test(
    has_rule( 'public', 'sometab', 'ins_me', 'whatever' ),
    true,
    'has_rule(schema, table, rule, desc)',
    'whatever',
    ''
);

SELECT * FROM check_test(
    has_rule( 'public', 'sometab', 'ins_me'::name ),
    true,
    'has_rule(schema, table, rule)',
    'Relation public.sometab should have rule ins_me',
    ''
);

SELECT * FROM check_test(
    has_rule( 'public', 'sometab', 'del_me', 'whatever' ),
    false,
    'has_rule(schema, table, rule, desc) fail',
    'whatever',
    ''
);

SELECT * FROM check_test(
    has_rule( 'sometab', 'ins_me', 'whatever' ),
    true,
    'has_rule(table, rule, desc)',
    'whatever',
    ''
);

SELECT * FROM check_test(
    has_rule( 'sometab', 'ins_me'::name ),
    true,
    'has_rule(table, rule)',
    'Relation sometab should have rule ins_me',
    ''
);

SELECT * FROM check_test(
    has_rule( 'sometab', 'del_me', 'whatever' ),
    false,
    'has_rule(table, rule, desc) fail',
    'whatever',
    ''
);

SELECT * FROM check_test(
    hasnt_rule( 'public', 'sometab', 'ins_me', 'whatever' ),
    false,
    'hasnt_rule(schema, table, rule, desc)',
    'whatever',
    ''
);

SELECT * FROM check_test(
    hasnt_rule( 'public', 'sometab', 'ins_me'::name ),
    false,
    'hasnt_rule(schema, table, rule)',
    'Relation public.sometab should not have rule ins_me',
    ''
);

SELECT * FROM check_test(
    hasnt_rule( 'public', 'sometab', 'del_me', 'whatever' ),
    true,
    'hasnt_rule(schema, table, rule, desc) fail',
    'whatever',
    ''
);

SELECT * FROM check_test(
    hasnt_rule( 'sometab', 'ins_me', 'whatever' ),
    false,
    'hasnt_rule(table, rule, desc)',
    'whatever',
    ''
);

SELECT * FROM check_test(
    hasnt_rule( 'sometab', 'ins_me'::name ),
    false,
    'hasnt_rule(table, rule)',
    'Relation sometab should not have rule ins_me',
    ''
);

SELECT * FROM check_test(
    hasnt_rule( 'sometab', 'del_me', 'whatever' ),
    true,
    'hasnt_rule(table, rule, desc) fail',
    'whatever',
    ''
);

/****************************************************************************/
-- Test rule_is_instead().

SELECT * FROM check_test(
    rule_is_instead( 'public', 'toview', '_RETURN', 'whatever' ),
    true,
    'rule_is_instead(schema, table, rule, desc)',
    'whatever',
    ''
);

SELECT * FROM check_test(
    rule_is_instead( 'public', 'toview', '_RETURN'::name ),
    true,
    'rule_is_instead(schema, table, rule)',
    'Rule "_RETURN" on relation public.toview should be an INSTEAD rule',
    ''
);

SELECT * FROM check_test(
    rule_is_instead( 'public', 'sometab', 'ins_me', 'whatever' ),
    false,
    'rule_is_instead(schema, table, nothing rule, desc)',
    'whatever',
    ''
);

SELECT * FROM check_test(
    rule_is_instead( 'public', 'sometab', 'upd_me', 'whatever' ),
    false,
    'rule_is_instead(schema, table, also rule, desc)',
    'whatever',
    ''
);

SELECT * FROM check_test(
    rule_is_instead( 'toview', '_RETURN', 'whatever' ),
    true,
    'rule_is_instead(table, rule, desc)',
    'whatever',
    ''
);

SELECT * FROM check_test(
    rule_is_instead( 'toview', '_RETURN'::name ),
    true,
    'rule_is_instead(table, rule)',
    'Rule "_RETURN" on relation toview should be an INSTEAD rule',
    ''
);

SELECT * FROM check_test(
    rule_is_instead( 'sometab', 'ins_me', 'whatever' ),
    false,
    'rule_is_instead(table, nothing rule, desc)',
    'whatever',
    ''
);

SELECT * FROM check_test(
    rule_is_instead( 'sometab', 'upd_me', 'whatever' ),
    false,
    'rule_is_instead(table, also rule, desc)',
    'whatever',
    ''
);

-- Check failure diagnostics for non-existent rules.
SELECT * FROM check_test(
    rule_is_instead( 'public', 'sometab', 'blah', 'whatever' ),
    false,
    'rule_is_instead(schema, table, non-existent rule, desc)',
    'whatever',
    '    Rule blah does not exist'
);

SELECT * FROM check_test(
    rule_is_instead( 'sometab', 'blah', 'whatever' ),
    false,
    'rule_is_instead(table, non-existent rule, desc)',
    'whatever',
    '    Rule blah does not exist'
);

/****************************************************************************/
-- Finish the tests and clean up.
SELECT * FROM finish();
ROLLBACK;
