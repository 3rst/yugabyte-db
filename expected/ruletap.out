\unset ECHO
1..66
ok 1 - has_rule(schema, table, rule, desc) should pass
ok 2 - has_rule(schema, table, rule, desc) should have the proper description
ok 3 - has_rule(schema, table, rule, desc) should have the proper diagnostics
ok 4 - has_rule(schema, table, rule) should pass
ok 5 - has_rule(schema, table, rule) should have the proper description
ok 6 - has_rule(schema, table, rule) should have the proper diagnostics
ok 7 - has_rule(schema, table, rule, desc) fail should fail
ok 8 - has_rule(schema, table, rule, desc) fail should have the proper description
ok 9 - has_rule(schema, table, rule, desc) fail should have the proper diagnostics
ok 10 - has_rule(table, rule, desc) should pass
ok 11 - has_rule(table, rule, desc) should have the proper description
ok 12 - has_rule(table, rule, desc) should have the proper diagnostics
ok 13 - has_rule(table, rule) should pass
ok 14 - has_rule(table, rule) should have the proper description
ok 15 - has_rule(table, rule) should have the proper diagnostics
ok 16 - has_rule(table, rule, desc) fail should fail
ok 17 - has_rule(table, rule, desc) fail should have the proper description
ok 18 - has_rule(table, rule, desc) fail should have the proper diagnostics
ok 19 - hasnt_rule(schema, table, rule, desc) should fail
ok 20 - hasnt_rule(schema, table, rule, desc) should have the proper description
ok 21 - hasnt_rule(schema, table, rule, desc) should have the proper diagnostics
ok 22 - hasnt_rule(schema, table, rule) should fail
ok 23 - hasnt_rule(schema, table, rule) should have the proper description
ok 24 - hasnt_rule(schema, table, rule) should have the proper diagnostics
ok 25 - hasnt_rule(schema, table, rule, desc) fail should pass
ok 26 - hasnt_rule(schema, table, rule, desc) fail should have the proper description
ok 27 - hasnt_rule(schema, table, rule, desc) fail should have the proper diagnostics
ok 28 - hasnt_rule(table, rule, desc) should fail
ok 29 - hasnt_rule(table, rule, desc) should have the proper description
ok 30 - hasnt_rule(table, rule, desc) should have the proper diagnostics
ok 31 - hasnt_rule(table, rule) should fail
ok 32 - hasnt_rule(table, rule) should have the proper description
ok 33 - hasnt_rule(table, rule) should have the proper diagnostics
ok 34 - hasnt_rule(table, rule, desc) fail should pass
ok 35 - hasnt_rule(table, rule, desc) fail should have the proper description
ok 36 - hasnt_rule(table, rule, desc) fail should have the proper diagnostics
ok 37 - rule_is_instead(schema, table, rule, desc) should pass
ok 38 - rule_is_instead(schema, table, rule, desc) should have the proper description
ok 39 - rule_is_instead(schema, table, rule, desc) should have the proper diagnostics
ok 40 - rule_is_instead(schema, table, rule) should pass
ok 41 - rule_is_instead(schema, table, rule) should have the proper description
ok 42 - rule_is_instead(schema, table, rule) should have the proper diagnostics
ok 43 - rule_is_instead(schema, table, nothing rule, desc) should fail
ok 44 - rule_is_instead(schema, table, nothing rule, desc) should have the proper description
ok 45 - rule_is_instead(schema, table, nothing rule, desc) should have the proper diagnostics
ok 46 - rule_is_instead(schema, table, also rule, desc) should fail
ok 47 - rule_is_instead(schema, table, also rule, desc) should have the proper description
ok 48 - rule_is_instead(schema, table, also rule, desc) should have the proper diagnostics
ok 49 - rule_is_instead(table, rule, desc) should pass
ok 50 - rule_is_instead(table, rule, desc) should have the proper description
ok 51 - rule_is_instead(table, rule, desc) should have the proper diagnostics
ok 52 - rule_is_instead(table, rule) should pass
ok 53 - rule_is_instead(table, rule) should have the proper description
ok 54 - rule_is_instead(table, rule) should have the proper diagnostics
ok 55 - rule_is_instead(table, nothing rule, desc) should fail
ok 56 - rule_is_instead(table, nothing rule, desc) should have the proper description
ok 57 - rule_is_instead(table, nothing rule, desc) should have the proper diagnostics
ok 58 - rule_is_instead(table, also rule, desc) should fail
ok 59 - rule_is_instead(table, also rule, desc) should have the proper description
ok 60 - rule_is_instead(table, also rule, desc) should have the proper diagnostics
ok 61 - rule_is_instead(schema, table, non-existent rule, desc) should fail
ok 62 - rule_is_instead(schema, table, non-existent rule, desc) should have the proper description
ok 63 - rule_is_instead(schema, table, non-existent rule, desc) should have the proper diagnostics
ok 64 - rule_is_instead(table, non-existent rule, desc) should fail
ok 65 - rule_is_instead(table, non-existent rule, desc) should have the proper description
ok 66 - rule_is_instead(table, non-existent rule, desc) should have the proper diagnostics
